src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from collections import OrderedDict
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import ClassVar, List, Optional, Tuple
+
+import pytest
+
+from pants.engine.fs import EMPTY_DIRECTORY_DIGEST, PathGlobs, Snapshot
+from pants.engine.rules import UnionMembership, rule
+from pants.engine.selectors import Get
+from pants.engine.target import AsyncField, PrimitiveField, Target
+from pants.testutil.engine.util import MockGet, run_rule
+from pants.util.collections import ensure_str_list
+from pants.util.memo import memoized_property
+from pants.util.ordered_set import OrderedSet
+
+
+class HaskellGhcExtensions(PrimitiveField):
+    alias: ClassVar = "ghc_extensions"
+    raw_value: Optional[List[str]]
+
+    @memoized_property
+    def value(self) -> List[str]:
+        if self.raw_value is None:
+            return []
+        # Add some arbitrary validation to test that hydration works properly.
+        bad_extensions = [
+            extension for extension in self.raw_value if not extension.startswith("Ghc")
+        ]
+        if bad_extensions:
+            raise ValueError(
+                f"All elements of `{self.alias}` must be prefixed by `Ghc`. Received "
+                f"{bad_extensions}."
+            )
+        return self.raw_value
+
+
+class HaskellSources(AsyncField):
+    alias: ClassVar = "sources"
+    raw_value: Optional[List[str]]
+
+    def validate_pre_hydration(self) -> None:
+        ensure_str_list(self.raw_value)
+
+    def validate_post_hydration(self, result: Snapshot) -> None:
+        non_haskell_sources = [fp for fp in result.files if PurePath(fp).suffix != ".hs"]
+        if non_haskell_sources:
+            raise ValueError(
+                f"Received non-Haskell sources in {self.alias}: {non_haskell_sources}."
+            )
+
+
+@dataclass(frozen=True)
+class HaskellSourcesResult:
+    snapshot: Snapshot
+
+
+@rule
+async def hydrate_haskell_sources(sources: HaskellSources) -> HaskellSourcesResult:
+    sources.validate_pre_hydration()
+    result = await Get[Snapshot](PathGlobs("*.hs"))
+    sources.validate_post_hydration(result)
+    return HaskellSourcesResult(result)
+
+
+class HaskellTarget(Target):
+    alias: ClassVar = "haskell"
+    core_fields: ClassVar = (HaskellGhcExtensions, HaskellSources)
```

Comment:
This is all it will ever take to define a new target type! 3 lines of declarative code 🎉

Reply:
Nice!

So... while I recognize the desire to keep the Target declaration site very small, having a different pattern for "core" (better name might be "default"?) fields than for extension fields might actually make things more challenging to document and explain.

It's ok to roll with this for now though I think.

Reply:
> having a different pattern for "core" (better name might be "default"?) fields than for extension fields might actually make things more challenging to document and explain.

Generally, users will always want to access `self.field_types`. This is what everything like `Target.get()` and `Target.has_sources()` uses. So, plugin fields do behave as first-class citizens.

We only differentiate between core fields vs. plugin fields at the constructor time so that:
1) We have better debugging. You can see in `repr()` which fields were added by you.
2) We can declare `core_fields` as a class var, whereas `plugin_types` must be an instance var because we have to inspect `UnionRule` at runtime. It's generally easier to work with class vars so we should use that where possible.

--

Why not unify plugin and core fields to always be set up via `UnionMembership`? Originally, I did take this approach. See the first few commits. But, it resulted in horrible verbosity and violated the principle of proximity. 

Instead of `core_fields: ClassVar = (HaskellGhcExtensions, HaskellSources)` in the class definition, we'd need `[UnionRule(HaskellTargetField, HaskellGhcExtensions), UnionRule(HaskellTargetField, HaskellSources)`. 

There's not a very compelling reason to have that verbosity + reduced proximity.